tools/parity_scale_decoder/src/main.rs

```diff
@@ -0,0 +1,279 @@
+        Less,
+        Metadata,
+        MetadataLimits,
+        MintBox,
+        Mod,
+        Multiply,
+        Not,
+        NotPermittedFail,
+        Or,
+        Pair,
+        Payload,
+        Peer,
+        PeerEvent,
+        PeerEventFilter,
+        PeerFilter,
+        PeerId,
+        PendingTransactions,
+        PermissionToken,
+        PipelineEntityKind,
+        PipelineEvent,
+        PipelineEventFilter,
+        PipelineStatus,
+        QueryBox,
+        QueryRequest,
+        QueryResult,
+        RaiseTo,
+        RegisterBox,
+        RejectedTransaction,
+        RejectionReason,
+        RemoveKeyValueBox,
+        Repeats,
+        RevokeBox,
+        Role,
+        RoleEvent,
+        RoleEventFilter,
+        RoleFilter,
+        RoleId,
+        SequenceBox,
+        SetKeyValueBox,
+        SignatureCheckCondition,
+        SignedQueryRequest,
+        Subtract,
+        TimeEvent,
+        TimeEventFilter,
+        TimeInterval,
+        TimeSchedule,
+        Transaction,
+        TransactionRejectionReason,
+        TransactionValue,
+        TransferBox,
+        Trigger,
+        TriggerEvent,
+        TriggerEventFilter,
+        TriggerFilter,
+        TriggerId,
+        UnregisterBox,
+        UnsatisfiedSignatureConditionFail,
+        ValidTransaction,
+        VersionedPendingTransactions,
+        VersionedQueryResult,
+        VersionedRejectedTransaction,
+        VersionedSignedQueryRequest,
+        VersionedTransaction,
+        VersionedValidTransaction,
+        WasmExecutionFail,
+        Where,
+    )
+}
+
+fn decode<W: io::Write>(args: DecodeArgs, map: &DumpDecodedMap, writer: &mut W) -> Result<()> {
+    let bytes = fs::read(args.binary)?;
+
+    if let Some(type_id) = args.type_id {
+        return decode_by_type(map, &type_id, &bytes, writer);
+    }
+    decode_by_guess(map, &bytes, writer)
+}
+
+fn decode_by_type<W: io::Write>(
+    map: &DumpDecodedMap,
+    type_id: &str,
+    bytes: &[u8],
+    writer: &mut W,
+) -> Result<()> {
+    map.get(type_id).map_or_else(
+        || Err(eyre!("Unknown type: `{type_id}`")),
+        |dump_decoded| dump_decoded(bytes, writer),
+    )
+}
+
+fn decode_by_guess<W: io::Write>(map: &DumpDecodedMap, bytes: &[u8], writer: &mut W) -> Result<()> {
+    let count = map
+        .values()
+        .filter_map(|dump_decoded| {
+            dump_decoded(bytes, writer)
+                .ok()
+                .and_then(|_| writeln!(writer).ok())
+        })
+        .count();
+    match count {
+        0 => writeln!(writer, "No compatible types found"),
+        1 => writeln!(writer, "1 compatible type found"),
+        n => writeln!(writer, "{n} compatible types found"),
+    }
+    .map_err(Into::into)
+}
+
+fn list_types<W: io::Write>(map: &DumpDecodedMap, writer: &mut W) -> Result<()> {
+    for key in map.keys() {
+        writeln!(writer, "{key}")?;
+    }
+
+    match map.len() {
+        0 => writeln!(writer, "No type is supported"),
```

Comment:
Better yet, why not make the "\n" global and remove it from the other match arms.

Reply:
There is no need for `\n` here. In that case we don't have any output before, so we don't need to separate this output. Otherwise it would be like this:

```bash
$ cargo run --bin parity_scale_decoder -- list-types

No type is supported
$
```

But we want:

```bash
$ cargo run --bin parity_scale_decoder -- list-types
No type is supported
$
```

Reply:
I would suggest moving the new lines to trail the previous output in that case. It's not at all obvious.